### ISSUE TITLE
win_hotfix does not support the newer combined cumulative update and service stack update packages

### ISSUE THREAD

**Issue by arylatt** [ansible-collections/community.windows#284]
##### SUMMARY

The `win_hotfix` module does not support installing MSU files that contain a Service Stack Update (SSU) and Latest Cumulative Update (LCU) within the same archive. For example, [KB5005033](https://support.microsoft.com/en-us/topic/august-10-2021-kb5005033-os-builds-19041-1165-19042-1165-and-19043-1165-b4c77d08-435a-4833-b9f7-e092372079a4).

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

[`win_hotfix`](https://github.com/ansible-collections/community.windows/blob/main/plugins/modules/win_hotfix.ps1)

##### ANSIBLE VERSION

```paste below
ansible [core 2.11.4]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /root/.local/lib/python3.9/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /root/.local/bin/ansible
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
  jinja version = 3.0.1
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /root/.local/lib/python3.9/site-packages/ansible_collections
Collection        Version
----------------- -------
community.windows 1.6.0
```

##### CONFIGURATION

Not available as I am running Ansible through Packer.

##### OS / ENVIRONMENT

**Ansible controller:**
```
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
```

**Target Host:**
Windows Server 2019 Datacenter edition (latest Azure image)

##### STEPS TO REPRODUCE

```yaml
tasks:
  - name: "Download MSU for KB5005033"
    win_get_url:
      dest: "C:\\Windows\\Temp\\KB5005033.msu"
      url: "http://download.windowsupdate.com/d/msdownload/update/software/secu/2021/08/windows10.0-kb5005033-x64_ebab415d7a65f0b33f93e9a30875d74baa8930a7.msu"

  - name: "Install MSU for KB5005033"
    win_hotfix:
      source: "C:\\Windows\\Temp\\KB5005033.msu"
      state: present
```

##### EXPECTED RESULTS

`win_hotfix` installs the specified MSU.

##### ACTUAL RESULTS

`win_hotfix` errors:

```paste below
    azure-arm.wsp: TASK [Install MSU for KB5005033] ***********************************************
    azure-arm.wsp: task path: /mnt/c/Users/andrewry/Downloads/packer/ansible/win_update.yml:8
    azure-arm.wsp: redirecting (type: modules) ansible.builtin.win_hotfix to community.windows.win_hotfix
    azure-arm.wsp: Using module file /root/.local/lib/python3.9/site-packages/ansible_collections/community/windows/plugins/modules/win_hotfix.ps1
    azure-arm.wsp: Pipelining is enabled.
    azure-arm.wsp: <20.49.185.113> ESTABLISH WINRM CONNECTION FOR USER: packer on PORT 5986 TO 20.49.185.113
    azure-arm.wsp: EXEC (via pipeline wrapper)
    azure-arm.wsp: The full traceback is:
    azure-arm.wsp: Cannot convert 'System.Object[]' to the type 'System.String' required by parameter 'ChildPath'. Specified method is not supported.
    azure-arm.wsp: At line:93 char:58
    azure-arm.wsp: + ... _file = Join-Path -Path $extract_path -ChildPath $cab_source_filename
    azure-arm.wsp: +                                                      ~~~~~~~~~~~~~~~~~~~~
    azure-arm.wsp:     + CategoryInfo          : InvalidArgument: (:) [Join-Path], ParentContainsErrorRecordException
    azure-arm.wsp:     + FullyQualifiedErrorId : CannotConvertArgument,Microsoft.PowerShell.Commands.JoinPathCommand
    azure-arm.wsp:
    azure-arm.wsp: ScriptStackTrace:
    azure-arm.wsp: at Get-HotfixMetadataFromFile, <No file>: line 93
    azure-arm.wsp: at <ScriptBlock>, <No file>: line 202
    azure-arm.wsp: fatal: [default]: FAILED! => {
    azure-arm.wsp:     "changed": false,
    azure-arm.wsp:     "msg": "Unhandled exception while executing module: Cannot convert 'System.Object[]' to the type 'System.String' required by parameter 'ChildPath'. Specified method is not supported."
    azure-arm.wsp: }
```

The bug appears to be occurring because of how `win_hotfix`'s [`Get-HotfixMetadataFromFile`](https://github.com/ansible-collections/community.windows/blob/main/plugins/modules/win_hotfix.ps1#L83) function expects only a single package to be listed in the XML file inside the MSU:

[`win_hotfix.ps1` L#91](https://github.com/ansible-collections/community.windows/blob/main/plugins/modules/win_hotfix.ps1#L91)
```
$cab_source_filename = $xml.unattend.servicing.package.source.GetAttribute("location")
```

However a combined SSU/LCU hotfix contains two packages:

```xml
<?xml version="1.0" encoding="utf-8"?>
<unattend xmlns="urn:schemas-microsoft-com:unattend">
  <servicing>
    <package action="install">
      <assemblyIdentity name="Package_for_ServicingStack" version="19041.1161.1.1" language="neutral" processorArchitecture="amd64" publicKeyToken="31bf3856ad364e35" xmlns="urn:schemas-microsoft-com:asm.v3" />
      <source location="%configsetroot%\SSU-19041.1161-x64.cab" />
    </package>
    <package action="install">
      <assemblyIdentity name="Multiple_Packages" version="0.0.0.0" language="neutral" processorArchitecture="" publicKeyToken="" />
      <source location="%configsetroot%\Windows10.0-KB5005033-x64.cab" />
    </package>
  </servicing>
</unattend>
```

**Comment by fsteve800** [comment]
```
$source=$xml.unattend.servicing.package
$source | foreach-object {$_.source.location}
```
This should populate the location from xml

**Comment by jborean93** [comment]
A PR is definitely welcome to fix this problem. I'm unsure whether all the cab files would need to be installed and not just the first 1 though. Might just need to loop this particular process to install each `location` of the CAB files but I'm not really familiar with the format of cumulative updates.

**Comment by fsteve800** [comment]
Thank you Jordan for merging, am not yet able to use it though in Tower have informed people to update the collection.

Thanks and regards,

Steven
9869579335
________________________________
From: Jordan Borean ***@***.***>
Sent: Wednesday, April 20, 2022 12:09 PM
To: ansible-collections/community.windows ***@***.***>
Cc: Stevenson Fernandes ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [ansible-collections/community.windows] win_hotfix does not support the newer combined cumulative update and service stack update packages (#284)

Closed #284 via #386. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: ***@***.*** ‍ ‍ ‍ ‍ ‍ ‍ ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
ZjQcmQRYFpfptBannerEnd

Closed #284<https://github.com/ansible-collections/community.windows/issues/284> via #386<https://github.com/ansible-collections/community.windows/pull/386>.
